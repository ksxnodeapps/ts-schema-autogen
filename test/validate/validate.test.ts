import { ok, err } from '@tsfun/result'

import {
  Config,
  Instruction,
  OutputDescriptor,
  SymbolInstruction,
  ValidatorFactory
} from '@ts-schema-autogen/validate'

describe('validateConfig', () => {
  function validate (instance: unknown) {
    const factory = new ValidatorFactory()
    const result = factory.validateConfig(instance)
    return { factory, result }
  }

  describe('positive', () => {
    function setup () {
      const instance: Config = {
        instruction: {
          symbol: 'Foo',
          output: 'output'
        }
      }
      return { instance, ...validate(instance) }
    }

    it('returns an Ok', () => {
      const { instance, result } = setup()
      expect(result).toEqual(ok(instance))
    })

    it('returns an object that has value being input instance', () => {
      const { instance, result } = setup()
      expect(result.value).toBe(instance)
    })
  })

  describe('negative', () => {
    function setup () {
      const instance = 'Not Valid'
      return { instance, ...validate(instance) }
    }

    it('returns an Err', () => {
      const { result } = setup()
      expect(result).toEqual(err(expect.anything()))
    })

    it('matches snapshot', () => {
      const { result } = setup()
      expect(result).toMatchSnapshot({
        error: [{ schema: expect.any(Object) }]
      })
    })
  })
})

describe('validateInstruction', () => {
  function validate (instance: unknown) {
    const factory = new ValidatorFactory()
    const result = factory.validateInstruction(instance)
    return { factory, result }
  }

  describe('positive', () => {
    function setup () {
      const instance: Instruction = {
        symbol: 'Foo',
        output: 'output'
      }
      return { instance, ...validate(instance) }
    }

    it('returns an Ok', () => {
      const { instance, result } = setup()
      expect(result).toEqual(ok(instance))
    })

    it('returns an object that has value being input instance', () => {
      const { instance, result } = setup()
      expect(result.value).toBe(instance)
    })
  })

  describe('negative', () => {
    function setup () {
      const instance = 'Not Valid'
      return { instance, ...validate(instance) }
    }

    it('returns an Err', () => {
      const { result } = setup()
      expect(result).toEqual(err(expect.anything()))
    })

    it('matches snapshot', () => {
      const { result } = setup()
      expect(result).toMatchSnapshot({
        error: [{ schema: expect.any(Object) }]
      })
    })
  })
})

describe('validateOutputDescriptor', () => {
  function validate (instance: unknown) {
    const factory = new ValidatorFactory()
    const result = factory.validateOutputDescriptor(instance)
    return { factory, result }
  }

  describe('positive', () => {
    function setup () {
      const instance: OutputDescriptor = {
        filename: 'output'
      }
      return { instance, ...validate(instance) }
    }

    it('returns an Ok', () => {
      const { instance, result } = setup()
      expect(result).toEqual(ok(instance))
    })

    it('returns an object that has value being input instance', () => {
      const { instance, result } = setup()
      expect(result.value).toBe(instance)
    })
  })

  describe('negative', () => {
    function setup () {
      const instance = 'Not Valid'
      return { instance, ...validate(instance) }
    }

    it('returns an Err', () => {
      const { result } = setup()
      expect(result).toEqual(err(expect.anything()))
    })

    it('matches snapshot', () => {
      const { result } = setup()
      expect(result).toMatchSnapshot({
        error: [{ schema: expect.any(Object) }]
      })
    })
  })
})

describe('validateSymbolInstruction', () => {
  function validate (instance: unknown) {
    const factory = new ValidatorFactory()
    const result = factory.validateSymbolInstruction(instance)
    return { factory, result }
  }

  describe('positive', () => {
    function setup () {
      const instance: SymbolInstruction = {
        output: 'output',
        symbol: 'Foo'
      }
      return { instance, ...validate(instance) }
    }

    it('returns an Ok', () => {
      const { instance, result } = setup()
      expect(result).toEqual(ok(instance))
    })

    it('returns an object that has value being input instance', () => {
      const { instance, result } = setup()
      expect(result.value).toBe(instance)
    })
  })

  describe('negative', () => {
    function setup () {
      const instance = 'Not Valid'
      return { instance, ...validate(instance) }
    }

    it('returns an Err', () => {
      const { result } = setup()
      expect(result).toEqual(err(expect.anything()))
    })

    it('matches snapshot', () => {
      const { result } = setup()
      expect(result).toMatchSnapshot({
        error: [{ schema: expect.any(Object) }]
      })
    })
  })
})
