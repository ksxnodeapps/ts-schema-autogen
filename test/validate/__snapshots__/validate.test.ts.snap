// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateConfig negative matches snapshot 1`] = `
Object {
  "error": Array [
    Object {
      "argument": Array [
        "object",
      ],
      "instance": "Not Valid",
      "message": "is not of a type(s) object",
      "name": "type",
      "property": "instance",
      "schema": Any<Object>,
      "stack": "instance is not of a type(s) object",
    },
  ],
  "tag": false,
}
`;

exports[`validateInstruction negative matches snapshot 1`] = `
Object {
  "error": Array [
    Object {
      "argument": Array [
        "<#/definitions/MultiSymbolInstruction>",
        "<#/definitions/SingleSymbolInstruction>",
      ],
      "instance": "Not Valid",
      "message": "is not any of <#/definitions/MultiSymbolInstruction>,<#/definitions/SingleSymbolInstruction>",
      "name": "anyOf",
      "property": "instance",
      "schema": Any<Object>,
      "stack": "instance is not any of <#/definitions/MultiSymbolInstruction>,<#/definitions/SingleSymbolInstruction>",
    },
  ],
  "tag": false,
}
`;

exports[`validateOutputDescriptor negative matches snapshot 1`] = `
Object {
  "error": Array [
    Object {
      "argument": Array [
        "object",
      ],
      "instance": "Not Valid",
      "message": "is not of a type(s) object",
      "name": "type",
      "property": "instance",
      "schema": Any<Object>,
      "stack": "instance is not of a type(s) object",
    },
  ],
  "tag": false,
}
`;

exports[`validateSymbolInstruction negative matches snapshot 1`] = `
Object {
  "error": Array [
    Object {
      "argument": Array [
        "object",
      ],
      "instance": "Not Valid",
      "message": "is not of a type(s) object",
      "name": "type",
      "property": "instance",
      "schema": Any<Object>,
      "stack": "instance is not of a type(s) object",
    },
  ],
  "tag": false,
}
`;
